Description: When unlocking, if we can't get the grab, try to set focus on the desired window & try again.
Author: Ray Strode <rstrode@redhat.com>
Bug-Upstream: https://bugzilla.redhat.com/show_bug.cgi?id=905935
Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/+source/gnome-screensaver/+bug/1175717

Index: b/src/gs-grab-x11.c
===================================================================
--- a/src/gs-grab-x11.c
+++ b/src/gs-grab-x11.c
@@ -377,7 +377,7 @@
 }
 
 static void
-gs_grab_nuke_focus (void)
+gs_grab_move_focus (GdkWindow *window)
 {
         Window focus = 0;
         int    rev = 0;
@@ -388,7 +388,7 @@
 
         XGetInputFocus (GDK_DISPLAY_XDISPLAY (gdk_display_get_default ()), &focus, &rev);
 
-        XSetInputFocus (GDK_DISPLAY_XDISPLAY (gdk_display_get_default ()), None, RevertToNone, CurrentTime);
+        XSetInputFocus (GDK_DISPLAY_XDISPLAY (gdk_display_get_default ()), GDK_WINDOW_XID (window), RevertToNone, CurrentTime);
 
         gdk_error_trap_pop_ignored ();
 }
@@ -469,7 +469,7 @@
         if (kstatus != GDK_GRAB_SUCCESS) {
                 if (!focus_fuckus) {
                         focus_fuckus = TRUE;
-                        gs_grab_nuke_focus ();
+                        gs_grab_move_focus (window);
                         goto AGAIN;
                 }
         }
